File: Demos/CustomViews/app/src/main/java/com/example/customviews/MainActivity.java
package com.example.customviews;

import androidx.appcompat.app.AppCompatActivity;

import android.graphics.Color;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    private UnderlinedTextView myUnderlinedTextView;
    private Button btnChangeText;
    private Button btnChangeColorSize;

    private boolean isTextOriginal = true;
    private boolean isStyleOriginal = true;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        myUnderlinedTextView = findViewById(R.id.my_underlined_text_view);
        btnChangeText = findViewById(R.id.btn_change_text);
        btnChangeColorSize = findViewById(R.id.btn_change_color_size);

        // 设置初始文本
        myUnderlinedTextView.setText("小米的新价值观，以“真诚热爱”为核心，强调用户至上的理念。其中，“和用户交朋友”的理念表明小米将更加真诚地对待用户，赢得用户的信任，并将用户放在核心地位，不仅仅关注他们的需求，还将他们视为最重要的合作伙伴。\n" +
                "\n" +
                "另外，“工程师思维”标志着小米将继续坚持技术创新，敢于探索，不断推出创新的产品。这一理念还强调了技术和质量对小米的重要性，相信技术创新和卓越质量将助力小米不断壮大。\n" +
                "\n" +
                "“主人翁精神”的价值观意味着每个小米员工都将积极参与公司的使命和愿景，肩负起荣誉和责任，坚决不为短期目标而牺牲公司的长期价值。这将有助于保持公司的可持续增长。\n" +
                "\n" +
                "“信任第一”强调了诚信的重要性，无论是对待用户、同事、股东还是合作伙伴，小米都将始终秉承诚信原则，进行坦诚沟通。\n" +
                "\n" +
                "“共创共识”将鼓励团队内的合作和民主决策，确保每个员工都有权参与重要决策的制定，并将决策结果充分沟通，保持团队的一致性。\n" +
                "\n" +
                "“结果导向”的理念则明确强调了数据和成果的重要性，小米将注重最终结果，同时也重视里程碑的达");


        btnChangeText.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (isTextOriginal) {
                    myUnderlinedTextView.setText("这是新的短文本内容，用于演示文本变化和下划线自适应。自定义 View 的能力非常强大！\n\n您可以通过属性动画来进一步美化这些效果。");
                } else {
                    myUnderlinedTextView.setText("小米的新价值观，以“真诚热爱”为核心，强调用户至上的理念。其中，“和用户交朋友”的理念表明小米将更加真诚地对待用户，赢得用户的信任，并将用户放在核心地位，不仅仅关注他们的需求，还将他们视为最重要的合作伙伴。\n" +
                            "\n" +
                            "另外，“工程师思维”标志着小米将继续坚持技术创新，敢于探索，不断推出创新的产品。这一理念还强调了技术和质量对小米的重要性，相信技术创新和卓越质量将助力小米不断壮大。\n" +
                            "\n" +
                            "“主人翁精神”的价值观意味着每个小米员工都将积极参与公司的使命和愿景，肩负起荣誉和责任，坚决不为短期目标而牺牲公司的长期价值。这将有助于保持公司的可持续增长。\n" +
                            "\n" +
                            "“信任第一”强调了诚信的重要性，无论是对待用户、同事、股东还是合作伙伴，小米都将始终秉承诚信原则，进行坦诚沟通。\n" +
                            "\n" +
                            "“共创共识”将鼓励团队内的合作和民主决策，确保每个员工都有权参与重要决策的制定，并将决策结果充分沟通，保持团队的一致性。\n" +
                            "\n" +
                            "“结果导向”的理念则明确强调了数据和成果的重要性，小米将注重最终结果，同时也重视里程碑的达");
                }
                isTextOriginal = !isTextOriginal;
                Toast.makeText(MainActivity.this, "文本内容已切换", Toast.LENGTH_SHORT).show();
            }
        });

        btnChangeColorSize.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (isStyleOriginal) {
                    myUnderlinedTextView.setTextColor(Color.BLUE);
                    myUnderlinedTextView.setTextSize(20f);
                    myUnderlinedTextView.setUnderlineColor(Color.RED);
                    myUnderlinedTextView.setUnderlineThickness(2.5f);
                    myUnderlinedTextView.setUnderlineOffset(5f);
                } else {
                    myUnderlinedTextView.setTextColor(Color.BLACK);
                    myUnderlinedTextView.setTextSize(16f);
                    myUnderlinedTextView.setUnderlineColor(Color.GRAY);
                    myUnderlinedTextView.setUnderlineThickness(1.5f);
                    myUnderlinedTextView.setUnderlineOffset(3f);
                }
                isStyleOriginal = !isStyleOriginal;
                Toast.makeText(MainActivity.this, "样式已切换", Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: Day11_KotlinRecyclerView/app/src/main/java/com/example/day11_kotlinrecyclerview/MainActivity.kt
package com.example.day11_kotlinrecyclerview

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts

class MainActivity : AppCompatActivity() {

    private lateinit var recyclerView: RecyclerView
    private lateinit var myListAdapter: MyListAdapter
    private val dataList = mutableListOf<ListItem>()

    private lateinit var detailActivityLauncher: ActivityResultLauncher<Intent>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        recyclerView = findViewById(R.id.recyclerView)

        detailActivityLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            // 处理从 DetailActivity 返回的结果
            if (result.resultCode == Activity.RESULT_OK) {
                // 使用 getParcelableExtra<ListItem>() 获取返回的 ListItem
                // 由于 isLiked 现在在 data class 的主构造函数中，Parcelize 会正确处理它
                val updatedItem = result.data?.getParcelableExtra<ListItem>(DetailActivity.EXTRA_LIST_ITEM)
                updatedItem?.let { item ->
                    // 找到被更新的项在 dataList 中的索引
                    val index = dataList.indexOfFirst { it.id == item.id }
                    if (index != -1) {
                        // 直接用返回的更新后的项替换 dataList 中对应的旧项
                        // 此时 item 的 isLiked 状态已经是 DetailActivity 中修改后的最新状态
                        dataList[index] = item
                        // 提交新的列表到 ListAdapter，ListAdapter 会自动计算差异并更新 UI
                        myListAdapter.submitList(dataList.toMutableList()) // 提交一个副本
                        Toast.makeText(this, "列表项 ${item.id} 状态已更新并刷新列表", Toast.LENGTH_SHORT).show()
                    }
                }
            }
        }

        // 初始化适配器，传入两个回调：列表项点击（跳转详情）和点赞按钮点击
        myListAdapter = MyListAdapter(
            onItemClicked = { clickedItem ->
                // 跳转详情页的逻辑保持不变
                val intent = Intent(this, DetailActivity::class.java).apply {
                    // 传递被点击的 ListItem。由于 isLiked 现在是主构造函数的一部分，它会被正确序列化。
                    putExtra(DetailActivity.EXTRA_LIST_ITEM, clickedItem)
                }
                detailActivityLauncher.launch(intent)
            },
            onLikeClicked = { likedItem -> // 处理列表页的点赞按钮点击回调
                // 找到被点赞的项在 dataList 中的索引
                val index = dataList.indexOfFirst { it.id == likedItem.id }
                if (index != -1) {
                    // 创建一个新的 ListItem 实例，并复制旧实例的内容，只改变 isLiked 状态
                    // 由于 isLiked 已经在 data class 的主构造函数中，直接使用 copy(isLiked = ...) 即可
                    val updatedItem = when (likedItem) {
                        is ListItem.TextItem -> likedItem.copy(isLiked = !likedItem.isLiked)
                        is ListItem.ImageItem -> likedItem.copy(isLiked = !likedItem.isLiked)
                    }
                    // 更新 dataList 中对应位置的项
                    dataList[index] = updatedItem
                    // 提交新的列表到 ListAdapter，ListAdapter 会自动计算差异并更新 UI
                    myListAdapter.submitList(dataList.toMutableList()) // 提交一个副本
                    Toast.makeText(this, "列表项 ${likedItem.id} 点赞状态: ${if (updatedItem.isLiked) "已点赞" else "待点赞"}", Toast.LENGTH_SHORT).show()
                }
            }
        )

        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = myListAdapter

        prepareData()
        myListAdapter.submitList(dataList)
    }

    /**
     * 准备初始数据列表，包含文本项和图片项。
     */
    private fun prepareData() {
        // 添加文本项
        for (i in 0 until 10) {
            val longText = "这是第 ${i + 1} 个文本内容。这是一个更长的文本示例，用于在详情页展示更多的信息。我们可以添加多段文字，来模拟一篇短文章，以便更好地测试滚动视图和布局效果。例如，一段关于Android开发的介绍，或者一些随机的Lorem Ipsum文本。这个文本会很长，长到足以在屏幕上滚动，从而验证我们的UI布局是否合理。".repeat(2 + i % 3) // 重复几次，让文本更长
            dataList.add(ListItem.TextItem(id = "text_$i", textContent = longText))
        }
        // 添加图片项
        val imageUrls = listOf(
            "https://picsum.photos/id/10/300/200",
            "https://picsum.photos/id/20/300/200",
            "https://picsum.photos/id/30/300/200",
            "https://picsum.photos/id/40/300/200",
            "https://picsum.photos/id/50/300/200"
        )
        for (i in 0 until 5) {
            dataList.add(ListItem.ImageItem(id = "image_$i", imageUrl = imageUrls[i % imageUrls.size]))
        }
        // 随机打乱顺序，使文本和图片项交错显示
        dataList.shuffle()
    }
}
